### ISSUE TITLE
How to differentiate between REST and POST requests of Gutenberg inside "save_post"?

### ISSUE THREAD

**Issue by manzoorwanijk** [WordPress/gutenberg#12903]
As we know that Gutenberg sends two requests when we hit the Publish/Update button:

1. Request to WP REST API to create/update the post
2. Request to `post.php` to save/submit the metaboxes

`save_post` is triggered by both the requests. How do I know inside the `save_post` callback that the post is created/updated via Gutenberg, preferably in the second request? This is to avoid the double effect of the hook.

I know some would suggest to add some meta value in the first request and use it to short-circuit the second callback or use a transient in a similar way. Filling up the meta table with useless data is not a good idea and transients don't look to be a better solution IMHO.

Is there a consistent query argument or a header that can be relied upon? I couldn't find any such thing in the two requests.

**Comment by manzoorwanijk** [comment]
Or may be if it's possible to add some parameter to the REST request so as to use it in `save_post` or `rest_insert_{$post_type}` callback. I checked the JS document for Gutenberg but couldn't find it.

**Comment by TimothyBJacobs** [comment]
What are you doing on the `save_post` hook?

If you only want to do something on REST requests, you could use the `rest_after_insert_` hooks.

**Comment by manzoorwanijk** [comment]
@TimothyBJacobs thank you for your contribution to WP REST API - `rest_after_insert_` hooks :)

Let me explain:
My plugin uses `save_post` to share the posts to Social Media depending on the user preferences - instantly or with some delay. The user can override the default preferences for a post on Post Edit Screen, currently via metaboxes.
Now there are many ways a post can be published/updated:
1. Using Classic Editor - allows to override the settings via metaboxes
2. Using Gutenberg:
    (i) WP REST API request - no metabox data available to `save_post`
    (ii) `POST` request tp `post.php` - metabox data available
3. Using the plugins like Jetpack Email to Post or Scrapping plugins - no metaboxes available
4. Using WP REST API (non-Gutenberg) - no metaboxes available

The main problem here is that how can I differentiate between 2(i) and 4.
A user may sometimes select not to share the post using the metabox options, but that selected option won't be available to the request 2(i). Is there a way I can add my own field to the REST request sent in 2(i)? That will solve my problem.
Thanks

**Comment by n7studios** [comment]
@manzoorwanijk I had the same question for WordPress to Buffer and WordPress to Hootsuite.

Here's the current solution I came up with, abstracted as a class that might help:
https://gist.github.com/n7studios/56fd05f19f5da26f19f6da0ccb57b144

**Comment by n7studios** [comment]
@manzoorwanijk Agreed.

I notice Posts through the Classic Editor define an `_edit_last` meta key, which doesn't appear to exist when Posts are edited through Gutenberg (both define the `_edit_lock` meta key, however).

Wondering if there's some way to use this to 'detect' whether the Post is truly Gutenberg or not?

Also worth inspecting the request object when Gutenberg + non-Gutenberg requests are made through the REST API, to see if there are any differences that distinguish whether the Post was created in Gutenberg or not.

**Comment by designsimply** [comment]
I'm afraid I don't know enough to answer your questions directly, but I did want to say that I noticed a recent PR at https://github.com/WordPress/gutenberg/pull/13718 which attempts to fix a preview race condition reported at https://github.com/WordPress/gutenberg/issues/12617 and the discussion about the two types of requests you mentioned at the start (`REST API` for the post and `post.php` for the metaboxes). I am wondering if anything in that discussion or PR might help.

I also know that Jetpack has some social media interactions and that plugin is open source and wondered if it would be a good suggestion to say to check out that code to see if they have done anything similar that could help you figure out your case by chance?

**Comment by talldan** [comment]
As mentioned by @designsimply, it's worth pointing out the PR at #13718 which proposes to switch the order of the requests (there's no guarantees that will be merged.)

That may cause an issue for code that relies on the existing order, though it might also reduce some of the complexity since I think you'd be able to switch to just relying on the REST hook.

Also worth mentioning the second request only occurs when meta boxes are active.

I'll make sure to make the point that plugins might have some dependencies on the save order in #13718.

**Comment by mboynes** [comment]
In encountering this issue in the wild, I believe this to be a fairly significant problem. Recapping the important bits:

* Gutenberg saves data via one or two requests; the second save request only fires if there are meta boxes, so a plugin cannot depend on this request happening.
* If there are going to be two requests, the plugin can't necessarily depend on the first request, because the first doesn't contain all the data for the post.
* A plugin can't determine in the first request that there will ever be a second request.

If a plugin needs to run some process after a post is updated, for instance sync the "final" post data to some external resource, there's no longer a reliable way to do that. It used to be that one could pretty heavily rely on `save_post` at a very high priority number, or `wp_insert_post`, to ensure that all post data (with relatively rare exception) has been saved.

**Comment by youknowriad** [comment]
I believe there's a solution proposed here https://gist.github.com/n7studios/56fd05f19f5da26f19f6da0ccb57b144

And Gutenberg will always use two requests if you have meta boxes but there's no other way to support meta boxes otherwise. The meta boxes API is not deterministic enough to be able to run its saving hooks on the REST API call.

**Comment by mboynes** [comment]
This should not be closed. The proposed solution is not a sufficient workaround. Please review my comment above, this is a problem without a viable option as-is.

At the most basic level, gutenberg should include a flag in the first request that there is going to be a second request.

**Comment by youknowriad** [comment]
is it possible to check for existence of metaboxes on the server to figure that out?

**Comment by mboynes** [comment]
That would be a lot of work and I don’t think it would be reliable. Gutenberg knows if it would be sending a follow up request or not, right? Why not announce that intent?

**Comment by youknowriad** [comment]
Let's reopen to reconsider that.

**Comment by manzoorwanijk** [comment]
Hey guys, sorry for not being able to reply earlier. I have come up with a hack that I use in my plugin and it has worked well so far. I have used `PluginPostStatusInfo` component, you may use whatever you want.

```jsx

const {
	element: {
		createElement,
		useEffect,
	},
	plugins: {
		registerPlugin,
	},
	editPost: {
		PluginPostStatusInfo,
	},
	data: {
		withSelect,
		withDispatch,
	},
	compose: {
		compose,
	},
} = wp;

// The array/object key that will be sent with the REST request
const key = 'isGutenbergPost';

const PluginIsGutenbergPost = ( { setIsGutenbergPost, isDirty } ) => {
	useEffect( () => {
		setIsGutenbergPost();
	}, [ isDirty ] );
	return (
		<PluginPostStatusInfo>
			{ null }
		</PluginPostStatusInfo>
	);
};

const EnhancedIsGutenbergPost = compose( [
	withSelect( ( select ) => {
		return {
			isDirty: select( 'core/editor' ).isEditedPostDirty(),
		};
	} ),
	withDispatch( ( dispatch, _, { select } ) => {
		return {
			setIsGutenbergPost: () => {
				const isDirty = select( 'core/editor' ).isEditedPostDirty();
				const isGBPost = select( 'core/editor' ).getEditedPostAttribute( key ) || false;
				if ( ! isGBPost && isDirty ) {
					dispatch( 'core/editor' ).editPost( { [ key ]: true }, { undoIgnore: true } );
				}
			},
		};
	} ),
] )( PluginIsGutenbergPost );

registerPlugin( 'is-gutenberg-post', { render: EnhancedIsGutenbergPost } );
```

Then on the server side you can use `"rest_pre_insert_{$post_type}"` hook to set your global or static flags which you can then check in `save_post` or `wp_insert_post`

```php
add_filter( "rest_pre_insert_post", function ( $prepared_post, WP_REST_Request $request ) {
	// The key is same as we set in JS code
	$is_gutenberg_post = $request->get_param( 'isGutenbergPost' );

	// Update your flags here, like
	$GLOBALS['is_gutenberg_post'] = $is_gutenberg_post

	return $prepared_post;
}, 10, 2 );

add_action( 'wp_insert_post', function wp_insert_post( $post_id, $post ) {

	if ( ! empty( $GLOBALS['is_gutenberg_post'] ) ) {
		return;
	}

	// If we are here, it means the request is not from Gutenberg

}, 10, 2 );
```

**Comment by janboddez** [comment]
Running into this as well. (I could copy a bunch of code over to Gutenberg and basically maintain a PHP and a JS codebase for the classic and the block editor, but that would, well, lead to _other_ issues. So I'd much rather be able to detect Gutenberg requests _reliably_. `has_blocks()` etc. just doesn't cut it, in this case.)

Could we, in addition to `defined( 'REST_REQUEST' ) && REST_REQUEST` or `empty( $_REQUEST['meta-box-loader'] )` (either would be `true` for both the _first_ Gutenberg and another type of REST request) not also look for a referrer?

To see if a request came from the admin or an entirely different client? (Even if they can be spoofed and whatnot, it'd still be better than assuming any REST request is a Gutenberg request, no?)

Seems to work (unless of course some hosts unset refer(r)er URLs?):
```
if (
    defined( 'REST_REQUEST' ) && REST_REQUEST &&
    empty( $_REQUEST['meta-box-loader'] ) && // phpcs:ignore WordPress.Security.NonceVerification.Recommended
    0 === strpos( wp_get_referer(), admin_url() )
) {
    // This should run only the first time a post is saved from the block editor.
}
```
Oh, I'm using this inside a `transition_post_status` callback.

**Comment by janboddez** [comment]
Was also wondering if, in case the referrer is deemed unreliable, we couldn't check against the `wp_rest` nonce (which we'd need to fetch from the headers or `_wp_nonce` body params)?

**Comment by janboddez** [comment]
> Was also wondering if, in case the referrer is deemed unreliable, we couldn't check against the `wp_rest` nonce (which we'd need to fetch from the headers or `_wp_nonce` body params)?

Case anyone's interested, still, I've been using this bit of code to detect whether a request is coming from Gutenberg (but not, e.g., a mobile app, for which `REST_REQUEST` would be set as well).
```
function is_gutenberg() {
	if ( ! defined( 'REST_REQUEST' ) || ! REST_REQUEST ) {
		// Not a REST request.
		return false;
	}

	$nonce = null;

	if ( isset( $_REQUEST['_wpnonce'] ) ) {
		$nonce = $_REQUEST['_wpnonce'];
	} elseif ( isset( $_SERVER['HTTP_X_WP_NONCE'] ) ) {
		$nonce = $_SERVER['HTTP_X_WP_NONCE'];
	}

	if ( ! is_string( $nonce ) ) {
		return false;
	}

	// Check the nonce.
	return wp_verify_nonce( $nonce, 'wp_rest' );
}
```

In combination with the presence of `$_REQUEST['meta-box-loader']`, this could be used (at least, it's been working for me, for like a year) to tell the difference between the first and a potential second request originating from _the block editor_.

**Comment by Mamaduka** [comment]
I'm going to close this; similar technical help requests aren't actionable. Shared examples allow differentiation between request origins.

P.S. WP 6.5 also introduced `wp_is_serving_rest_request` helper functions.

**Comment by mboynes** [comment]
@Mamaduka I might be mistaken, but I don't believe that a viable long-term solution has been merged yet and I would ask that the issue remain open.

If you [review this comment](https://github.com/WordPress/gutenberg/issues/12903#issuecomment-479976609), the point isn't determining which is a POST and which is a REST request, **the problem is that GB sometimes saves a post via one request, sometimes via two requests, and a plugin or theme doesn't know which it will be**. The first request, as far as I'm aware, still doesn't indicate that there will be a second request or not.

If that plugin or theme wants to run some process (e.g. sync data to another site, or send content to Apple News, or run push notifications to mobile apps, or purge an edge cache, etc) after all of a post's data is done saving, there's no reliable way to know when that's happened.

If an update has been merged that solves that problem, it should be noted here for posterity.

**Comment by manzoorwanijk** [comment]
I agree that the issue still remains, and there is no deterministic way to know if there will be a second request.

**Comment by Mamaduka** [comment]
It might be better to move this request to Trac - https://core.trac.wordpress.org/newticket. The REST API and metabox backward compatibility issues are addressed there.

We can keep this issue open, but it's not actionable on the JS side.

**Comment by mboynes** [comment]
@Mamaduka the simplest solution might actually be on the JS side. As I understand it, and please correct me where I'm wrong, the editor dispatches the two save requests. In the first save request, it should theoretically know that it will be dispatching a second save request, is that right? If so, it could include in the first save request some additional data to indicate that it is a partial payload (e.g. a query string `?partial_payload=true|false`). The mere presence of this data, coupled with the enhancements you highlighted earlier, should be sufficient information for plugins and themes to know if the post is fully updated when `save_post` et al fire.

**Comment by manzoorwanijk** [comment]
It could be as simple as passing a query param to the first request based on [`hasMetaBoxes`](https://github.com/WordPress/gutenberg/blob/de9d8bf200d7a2b9c6bdfc4108c46147da5124bc/packages/edit-post/src/store/selectors.js#L434) selector.

**Comment by Mamaduka** [comment]
@mboynes, when legacy metaboxes are active in the editor, it will dispatch to requests:

1. The default REST API request for updating data. This is how the block editor always updates data.
2. Requests for metaboxes backward compatibility. You can check the URL by running `window._wpMetaBoxUrl` in DevTools console.

The second request already contains a special query argument that can be used to abort the `save_post` action.

## Screenshot

![Image](https://github.com/user-attachments/assets/1acca734-a31b-4dc5-aa77-f4449cae4e4e)

**Comment by mboynes** [comment]
@Mamaduka the problem is that the second request is not guaranteed, it only executes if there are legacy meta boxes. Therefore a plugin (for example) cannot determine when `save_post` fires during the first (REST) request that the post is fully updated or not. If the editor were to indicate in the REST request that a legacy meta box save is or is not forthcoming, then the plugin _would_ be able to determine if `save_post` fired during the REST request is final or if it needs to wait for `save_post` to execute during the POST request. Does that explanation make the issue more clear?

**Comment by Mamaduka** [comment]
@mboynes, so you would like to return early in the `save_post` action triggered by the REST API, instead of the backward compatibility one for metaboxes. Is that correct?

**Comment by mboynes** [comment]
Yes, if I understand what you mean, a callback would return early during the REST request if it could determine that a followup legacy metaboxes save were forthcoming. Here's what part of a `save_post` callback might look like if the REST request included `partial_payload=true|false`:

```php
add_action( 'save_post', function ( $post_id ) {
    if (
        wp_is_serving_rest_request()
        && isset( $_REQUEST['partial_payload'] )
        && 'true' === $_REQUEST['partial_payload']
    ) {
        // The post data is incomplete, using it here is premature.
        return;
    }

    // Otherwise, it is safe to assume the post data is complete and done being saved.
}, 1, 999 );
```
